### ISSUE TITLE
ssl.SSLError: [CRYPTO] unknown error (_ssl.c:3076) on Ubuntu Pro 20.04 LTS

### ISSUE THREAD

**Issue by mttjohnson-jf** [certbot/certbot#10044]
The latest version `3.0.0` of certbot is producing an error, where the previously installed version `2.11.0` was not. The error is produced even when just trying to output the version (`certbot --version`).

## My operating system is (include version):
Ubuntu 20.04.6 LTS (with Ubuntu Pro license and fips-updates enabled)

## I installed Certbot with (snap, OS package manager, pip, certbot-auto, etc):
snap

## I ran these commands:
```
certbot --version
certbot --help
certbot
```
...and they all produced this same output:
```
Traceback (most recent call last):
  File "/snap/certbot/4182/bin/certbot", line 5, in <module>
    from certbot.main import main
  File "/snap/certbot/4182/lib/python3.12/site-packages/certbot/main.py", line 6, in <module>
    from certbot._internal import main as internal_main
  File "/snap/certbot/4182/lib/python3.12/site-packages/certbot/_internal/main.py", line 23, in <module>
    from acme import client as acme_client
  File "/snap/certbot/4182/lib/python3.12/site-packages/acme/client.py", line 20, in <module>
    import requests
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/__init__.py", line 164, in <module>
    from .api import delete, get, head, options, patch, post, put, request
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/api.py", line 11, in <module>
    from . import sessions
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/sessions.py", line 15, in <module>
    from .adapters import HTTPAdapter
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/adapters.py", line 80, in <module>
    _preloaded_ssl_context = create_urllib3_context()
                             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/snap/certbot/4182/lib/python3.12/site-packages/urllib3/util/ssl_.py", line 299, in create_urllib3_context
    context = SSLContext(ssl_version)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/snap/certbot/4182/usr/lib/python3.12/ssl.py", line 438, in __new__
    self = _SSLContext.__new__(cls, protocol)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLError: [CRYPTO] unknown error (_ssl.c:3076)
```

## Certbot's behavior differed from what I expected because:
The latest version (`3.0.0`) of the snap that was just installed automatically, produces an error when running any `certbot` command.
```
$ sudo snap list certbot --all
Name     Version  Rev   Tracking       Publisher     Notes
certbot  2.11.0   3834  latest/stable  certbot-eff✓  disabled,classic
certbot  3.0.0    4182  latest/stable  certbot-eff✓  classic
```

I was able to temporarily workaround the issue by reverting the installed version:
```bash
sudo snap revert certbot
```
...or installing an older revision (version 2.11.0) on a new system
```bash
sudo snap refresh certbot --channel=latest/stable --revision 3834
```
...and then setting a 1 week hold on auto updating the snap:
```bash
sudo snap refresh --hold=168h certbot
```

After reverting the installed version back to 2.11.0 the `certbot --version` command output the version number as expected.
```
$ sudo snap revert certbot
Stop snap "certbot" services                                                                                                                       |
certbot reverted to 2.11.0
$ sudo snap refresh --hold=168h certbot
General refreshes of "certbot" held until 2024-11-14T14:30:17-05:00
$ sudo snap list certbot --all
Name     Version  Rev   Tracking       Publisher     Notes
certbot  2.11.0   3834  latest/stable  certbot-eff✓  classic,held
certbot  3.0.0    4182  latest/stable  certbot-eff✓  disabled,classic,held
$ certbot --version
certbot 2.11.0
```

###### Logs are stored in `/var/log/letsencrypt` by default. Feel free to redact domains, e-mail and IP addresses as you see fit.
No log data was written to `/var/log/letsencrypt/letsencrypt.log` when running `certbot --version`.

**Comment by DorianCoding** [comment]
Hello,

It seems the certificate used by LetsEncrypt is not recognized by your computer. Are certificates in `/etc/ssl/certs` up to date? Do you have more messages with the verbose mode?

**Comment by mttjohnson-jf** [comment]
The system has the latest (https://launchpad.net/ubuntu/focal/+source/ca-certificates) `ca-certificates` package installed:
```
$ apt list ca-certificates
Listing... Done
ca-certificates/focal-updates,focal-security,now 20240203~20.04.1 all [installed]
```

The `--verbose` flag (`certbot --verbose` or `certbot --verbose --version`) doesn't do anything different, no additional information, and no log data written.

If this is an issue with a certificate, do we know what specific certificate this is having issues with when I run the `certbot` command?

**Comment by DorianCoding** [comment]
Won't be of much help. It's probably due to problems between python 3.8 and python 3.12. Are you using python 3.12 `python3 -version` ?

**Comment by mttjohnson-jf** [comment]
The version of python on the system is just the default installed system version maintained by the OS distribution:
```
$ python3 --version
Python 3.8.10
$ apt list python3
Listing... Done
python3/focal,now 3.8.2-0ubuntu2 amd64 [installed]
```

Based on the path listed in the error message above (`/snap/certbot/4182/usr/lib/python3.12/ssl.py`) it looks like the snap bundles it's own version of Python 3.12 stuff that it uses for the `certbot` command.

**Comment by mttjohnson-jf** [comment]
I went ahead and spun up a separate VM running Ubuntu 20.04 LTS that doesn't have anything to do with Ubuntu Pro and FIPS mode, so it has the normal OpenSSL distribution installed:
```
openssl/focal-updates,focal-security,now 1.1.1f-1ubuntu2.23 amd64 [installed]
```

I checked the contents of `/etc/ssl/openssl.cnf` on both systems and they are the same, so that can be ruled out as a potential difference.

While this quick new stock Ubuntu 20.04 LTS VM (different hypervisor and platform) I setup is different than the one I have with issues above, I also went ahead and quickly installed `snapd` and then `certbot` (v3.0.0) via a snap, and it worked fine there. I also checked the Python version, and is also using the system default v3.8.10 on the new stock system. This leads me to suspect that it may be more related to some differences with the Ubuntu Pro and FIPS mode support that isn't playing well with certbot v3.0.0, but I'm not sure what is needed to further isolate the root cause here.

**Comment by DorianCoding** [comment]
> I went ahead and spun up a separate VM running Ubuntu 20.04 LTS that doesn't have anything to do with Ubuntu Pro and FIPS mode, so it has the normal OpenSSL distribution installed:

Yes probably. Maybe you can play with parameters of ssl on the python files of certbot to enable others algorithms.

**Comment by DorianCoding** [comment]
> Everything is working fine with certbot v2.11.0 on the same system, and the differences seems to be something changed or was introduced with certbot v3.0.0.

It is not certbot errors. It is an error on the requests package. You can see it on your error messages. Therefore the version used by requests does not match your OS configuration.

**Comment by bmw** [comment]
hey all :wave:

thanks for all the detailed information, debugging so far, and temporary workaround of using an older snap as described in the initial post

this is a tricky one. this error does happen when just importing the very common python library `requests` which you can see in the traceback. now whether whatever is conflicting with FIPS is due to changes in requests, urllib3, python, or openssl, i'm not currently sure. i also unfortunately do not currently have a ubuntu 20.04 pro FIPS machine available i can play with

a snap command that may help here is `sudo snap run --shell certbot` which will open a shell in the same environment that the certbot snap runs in (although remember to update your certbot snap to the latest version again before playing with this). from inside this shell, you should be able to easily reproduce the problem with a command like `python -c 'import requests'`. from there, we could play with python options like `python -v -c 'import requets'` or tools like `strace`. these will produce A LOT of output, but they may include something that could help us debug the problem

another option that would be much more work but may help if necessary is for someone/us to try to build a version of the snap using the `python3-dbg` package instead of `python3` and then try to recreate the error using that snap. `python3-dbg` is a debug build of python which should hopefully allow us to get more information about the error occurring in `/snap/certbot/4182/usr/lib/python3.12/ssl.py` than is currently being provided

**Comment by mttjohnson-jf** [comment]
I built a new GCP VM instance for testing some of this in better isolation and confirmed it only occurs on the Ubunto Pro instance when FIPS mode is enabled.

compute image project: `ubuntu-os-pro-cloud`
compute image family: `ubuntu-pro-2004-lts`

Once this instance is created I made sure the system was updated then installed `certbot`:
```bash
# update system after clean install
sudo apt update
sudo apt full-upgrade
sudo apt install snapd
sudo snap refresh

# Install certbot snap
sudo snap install --classic certbot
```

Certbot worked just fine up to this point on the Ubuntu Pro 20.04 LTS:
```
$ certbot --version
certbot 3.0.0
```

I then installed the `fips-updates`:
```bash
# Ubuntu Pro FIPS Mode
# NOTE: GCP service account must be active on compute instance to automatically activate pro license
sudo pro status
sudo pro enable fips-updates --assume-yes
# Rebooted server to complete the install
```

After the system rebooted I confirmed it produced the same error when running `certbot --version`, then ran the commands you suggested to try and get some additional output. I'm only presenting the stuff before the error the relates to the requests package, as that seemed to be what you're wanting to focus on:
```
sudo snap run --shell certbot
root@matt-temp-utility-pro:/home/mattjohnson# python -v -c 'import requests'
...
import 'requests.utils' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae172f470>
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/api.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/api.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/api.cpython-312.pyc'
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/sessions.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/sessions.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/sessions.cpython-312.pyc'
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/adapters.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/adapters.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/adapters.cpython-312.pyc'
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/auth.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/auth.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/auth.cpython-312.pyc'
import 'requests.auth' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae14fd130>
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/models.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/models.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/models.cpython-312.pyc'
# /snap/certbot/4182/usr/lib/python3.12/encodings/__pycache__/idna.cpython-312.pyc matches /snap/certbot/4182/usr/lib/python3.12/encodings/idna.py
# code object from '/snap/certbot/4182/usr/lib/python3.12/encodings/__pycache__/idna.cpython-312.pyc'
# /snap/certbot/4182/usr/lib/python3.12/__pycache__/stringprep.cpython-312.pyc matches /snap/certbot/4182/usr/lib/python3.12/stringprep.py
# code object from '/snap/certbot/4182/usr/lib/python3.12/__pycache__/stringprep.cpython-312.pyc'
import 'stringprep' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae14ff2f0>
import 'encodings.idna' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae14fed20>
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/hooks.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/hooks.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/hooks.cpython-312.pyc'
import 'requests.hooks' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae14fffe0>
# /snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/status_codes.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/requests/status_codes.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/requests/__pycache__/status_codes.cpython-312.pyc'
import 'requests.status_codes' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae12b81a0>
import 'requests.models' # <_frozen_importlib_external.SourceFileLoader object at 0x7faae14fdac0>
# /snap/certbot/4182/lib/python3.12/site-packages/urllib3/contrib/__pycache__/socks.cpython-312.pyc matches /snap/certbot/4182/lib/python3.12/site-packages/urllib3/contrib/socks.py
# code object from '/snap/certbot/4182/lib/python3.12/site-packages/urllib3/contrib/__pycache__/socks.cpython-312.pyc'
# destroy urllib3.contrib.socks
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/__init__.py", line 164, in <module>
    from .api import delete, get, head, options, patch, post, put, request
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/api.py", line 11, in <module>
    from . import sessions
  File "<frozen importlib._bootstrap>", line 1415, in _handle_fromlist
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/sessions.py", line 15, in <module>
    from .adapters import HTTPAdapter
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/adapters.py", line 80, in <module>
    _preloaded_ssl_context = create_urllib3_context()
                             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/snap/certbot/4182/lib/python3.12/site-packages/urllib3/util/ssl_.py", line 299, in create_urllib3_context
    context = SSLContext(ssl_version)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/snap/certbot/4182/usr/lib/python3.12/ssl.py", line 438, in __new__
    self = _SSLContext.__new__(cls, protocol)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLError: [CRYPTO] unknown error (_ssl.c:3076)
...
```

I'm not sure if that provides anything more useful to identify where the issue is at, but I'm open to suggestions on digging into this further.

**Comment by wgreenberg** [comment]
that's a super helpful datapoint. could you also run `openssl version` while within the FIPS-enabled certbot snap shell?

**Comment by mttjohnson-jf** [comment]
It doesn't really display anything different or unique:
```
$ sudo snap run --shell certbot
root@matt-temp-utility-pro:/home/mattjohnson# openssl version
OpenSSL 1.1.1f  31 Mar 2020
```

It doesn't really display anything related to the FIPS version that is installed, but I can see that the apt package displays that it's got the fips version installed:
```
$ apt list openssl
Listing... Done
openssl/focal-updates,now 1.1.1f-1ubuntu2.fips.23 amd64 [installed,automatic]
```

**Comment by DorianCoding** [comment]
If you import requests alone does it work?

**Comment by mttjohnson-jf** [comment]
If I'm not in the certbot snap shell and using the system python version it doesn't produce an error, but if I switch to certbot snap shell I get the error:
```
$ python3 --version
Python 3.8.10
$ python3 -c 'import requests'
$ snap run --shell certbot
$ python3 --version
Python 3.12.3
$ python3 -c 'import requests'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/__init__.py", line 164, in <module>
    from .api import delete, get, head, options, patch, post, put, request
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/api.py", line 11, in <module>
    from . import sessions
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/sessions.py", line 15, in <module>
    from .adapters import HTTPAdapter
  File "/snap/certbot/4182/lib/python3.12/site-packages/requests/adapters.py", line 80, in <module>
    _preloaded_ssl_context = create_urllib3_context()
                             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/snap/certbot/4182/lib/python3.12/site-packages/urllib3/util/ssl_.py", line 299, in create_urllib3_context
    context = SSLContext(ssl_version)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/snap/certbot/4182/usr/lib/python3.12/ssl.py", line 438, in __new__
    self = _SSLContext.__new__(cls, protocol)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLError: [CRYPTO] unknown error (_ssl.c:3076)
```

**Comment by DorianCoding** [comment]
Can you replace on ```snap/certbot/4182/lib/python3.12/site-packages/urllib3/util/ssl_.py``` line 299 ```context = SSLContext(ssl_version)``` by ```context = SSLContext()```

**Comment by wgreenberg** [comment]
> Can you replace on `snap/certbot/4182/lib/python3.12/site-packages/urllib3/util/ssl_.py` line 299 `context = SSLContext(ssl_version)` by `context = SSLContext()`

i suspect that won't work due to snap mounting `$SNAP/lib` as a readonly filesystem, but i think you're on the right track here

**Comment by mttjohnson-jf** [comment]
> Can you replace on snap/certbot/4182/lib/python3.12/site-packages/urllib3/util/ssl_.py line 299 context = SSLContext(ssl_version) by context = SSLContext()

Yeah, the files there are read only, and it doesn't let me modify it.

**Comment by wgreenberg** [comment]
@mttjohnson-jf sorry for the back and forth, but could you run this inside both the snap shell and host environment: `python -c "import ssl; print(ssl._ssl.OPENSSL_VERSION); print(ssl._ssl)"`

**Comment by wgreenberg** [comment]
ok actually i was able to get ahold of a FIPS-enabled Ubuntu Pro installation, and i'm able to reproduce this locally. it's looking like the snap's python 3.12's `ssl` module is built on OpenSSL 3.0.13, while the system's python 3.8.10's `ssl` module is built on `1.1.1f`, which reflects what's installed on the system. not sure at all why this is the case...

**Comment by DorianCoding** [comment]
> ok actually i was able to get ahold of a FIPS-enabled Ubuntu Pro installation, and i'm able to reproduce this locally

Same for me. Your bug is indeed confirmed.

However using pip I don't have this error... Neither with python3.8 nor python 3.12 with `requests-2.32.3`.

And on the file, here is the diff:

```
>     context = SSLContext(ssl_version)
271,305c301
<         # 'ssl_version' is deprecated and will be removed in the future.
<         else:
<             # Use 'ssl_minimum_version' and 'ssl_maximum_version' instead.
<             ssl_minimum_version = _SSL_VERSION_TO_TLS_VERSION.get(
<                 ssl_version, TLSVersion.MINIMUM_SUPPORTED
<             )
<             ssl_maximum_version = _SSL_VERSION_TO_TLS_VERSION.get(
<                 ssl_version, TLSVersion.MAXIMUM_SUPPORTED
<             )
< 
<             # This warning message is pushing users to use 'ssl_minimum_version'
<             # instead of both min/max. Best practice is to only set the minimum version and
<             # keep the maximum version to be it's default value: 'TLSVersion.MAXIMUM_SUPPORTED'
<             warnings.warn(
<                 "'ssl_version' option is deprecated and will be "
<                 "removed in urllib3 v2.1.0. Instead use 'ssl_minimum_version'",
<                 category=DeprecationWarning,
<                 stacklevel=2,
<             )
< 
<     # PROTOCOL_TLS is deprecated in Python 3.10 so we always use PROTOCOL_TLS_CLIENT
<     context = SSLContext(PROTOCOL_TLS_CLIENT)
< 
<     if ssl_minimum_version is not None:
<         context.minimum_version = ssl_minimum_version
<     else:  # Python <3.10 defaults to 'MINIMUM_SUPPORTED' so explicitly set TLSv1.2 here
<         context.minimum_version = TLSVersion.TLSv1_2
< 
<     if ssl_maximum_version is not None:
<         context.maximum_version = ssl_maximum_version
< 
<     # Unless we're given ciphers defer to either system ciphers in
<     # the case of OpenSSL 1.1.1+ or use our own secure default ciphers.
<     if ciphers:
<         context.set_ciphers(ciphers)

```

With this on certbot: 
```
Package            Version
------------------ ---------
acme               3.0.0
certbot            3.0.0
certbot-apache     3.0.0
certbot-nginx      3.0.0
certifi            2024.8.30
cffi               1.17.1
charset-normalizer 3.4.0
ConfigArgParse     1.7
configobj          5.0.9
cryptography       43.0.3
distro             1.9.0
idna               3.10
josepy             1.14.0
parsedatetime      2.6
pip                24.3.1
pycparser          2.22
pyOpenSSL          24.2.1
pyparsing          3.1.4
pyRFC3339          1.1
python-augeas      1.1.0
pytz               2024.2
**requests           2.32.3**
setuptools         75.3.0
**urllib3            1.26.20**
wheel              0.44.0
```

Urllib is completely out of date, should be version 2+.
Asking an update with pip:

```
Collecting urllib3
  Using cached urllib3-2.2.3-py3-none-any.whl.metadata (6.5 kB)
Using cached urllib3-2.2.3-py3-none-any.whl (126 kB)
Installing collected packages: urllib3
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.26.20
    Not uninstalling urllib3 at /snap/certbot/4182/lib/python3.12/site-packages, outside environment /root/python3-12
    Can't uninstall 'urllib3'. No files were found to uninstall.
Successfully installed urllib3-2.2.3
```
Unfortunately because of the read-only mode it does not solve the problem.

Actually the question therefore is: why urllib3 is installed with a downgraded version and what is responsible of that in the snap package?

**Comment by bmw** [comment]
> Urllib is completely out of date, should be version 2+.

i'm not sure this is true. [urllib3 1.26.x is still supported](https://urllib3.readthedocs.io/en/1.26.x/v2-roadmap.html#v1-26-x-security-and-bug-fixes), [1.26.20 came out less than 3 months ago](https://github.com/urllib3/urllib3/releases/tag/1.26.20), and [we have it pinned back because some of our other dependencies haven't migrated yet](https://github.com/certbot/certbot/blob/38fc7fcc4804d8429fbdeb990fe1bc2ddf255629/tools/pinning/current/pyproject.toml#L11-L12)

i think there's many components in the stack used in the snap and it's not clear to me where the problem lies yet

**Comment by kkeane** [comment]
I doubt it is related to the version of openssl. I see the exact same error on RedHat EL 9 and AlmaLinux 9, which have openssl 3.0.7, and a similar issue (probably the same, but a different stack dump) on RHEL 8, which uses openssl 1.1.

All seem to only happen with FIPS mode enabled.

**Comment by DorianCoding** [comment]
> All seem to only happen with FIPS mode enabled. Non-FIPS systems work fine.

Using pip everything works fine even with FIPS mode enabled.

**Comment by wgreenberg** [comment]
dug into this more, and i think i have a better idea of what's going wrong. for context, i'm testing this on an Ubuntu Pro 20.04 system w/ FIPS enabled i can reproduce the same error using only the python3.12 installation within the certbot snap's standard library:

```
root@willg-ubuntu:/home/willg# snap run --shell certbot
root@willg-ubuntu:/home/willg# python3 -c 'import ssl; ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/snap/certbot/x1/usr/lib/python3.12/ssl.py", line 438, in __new__
    self = _SSLContext.__new__(cls, protocol)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLError: [CRYPTO] unknown error (_ssl.c:3076)
```

stracing this showed that python's `ssl` module calls into the OpenSSL 3.0.13 compiled/distributed with python3.12. one of the changes introduced in OpenSSL 3.x was the concept of FIPS providers, and once OpenSSL 3.x determines the system's FIPS enabled (by checking `/proc/sys/crypto/fips_enabled`), it tries to load a provider at `/usr/lib/x86_64-linux-gnu/ossl-modules/fips.so`. But since Ubuntu Pro's FIPS-enabled OpenSSL is v1.1.1f, it doesn't use providers, and neither does the snap's non-FIPS enabled OpenSSL 3.0.13. so finally, OpenSSL bails with the incredibly unhelpful "unknown error".

tl;dr, python3.12 uses a newer version of OpenSSL that expects a FIPS provider if it detects FIPS is enabled, but since Ubuntu's FIPS system distributes the older OpenSSL version w/o providers, it fails.

this whole thing has been my first experience w/ FIPS systems, so if anyone has more context/ideas about how to handle this, please chime in! otherwise, i see two possible solutions here:
1. certbot distributes a FIPS provider with its snap, either in the base snap or as a plugin
2. certbot users on FIPS-enabled systems should install certbot via `pip`. this is somewhat of a setback since we're trying to focus on supporting snap, but considering [even Ubuntu's official FIPS docs suggest installing certbot via pip](https://ubuntu.com/blog/manage-fips-enabled-linux-machines-at-scale-with-landscape-23-03), maybe it's the most robust solution going forward

**Comment by kkeane** [comment]
I think there may still be part of the story missing.

If the problem was an architectural mismatch between OpenSSL 1.1.1 and 3.0, the problem shouldn't happen on RHEL 9 (or AlmaLinux 9), because those have OpenSSL 3.0 natively.

But the problem happens on RHEL 9 (as well as RHEL 8, which has OpenSSL 1.1.1)

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

**Comment by wgreenberg** [comment]
interesting, i don’t have access to a RHEL system for testing, but i’m curious if you can see if RHEL 9 has an OpenSSL provider file somewhere? and if so, could you strace the snapped certbot process and see if it’s being picked up?

**Comment by kkeane** [comment]
I'm not enough of an expert on openssl, but it appears that RHEL 9 uses a mix of engines and providers - apparently, the migration to providers won't be completed until RHEL 10. FIPS is supposed to be implemented as a provider, though.

I see these files. I'm assuming that ossl-modules are the providers.

/usr/lib64/engines-3
/usr/lib64/engines-3/afalg.so
/usr/lib64/engines-3/capi.so
/usr/lib64/engines-3/loader_attic.so
/usr/lib64/engines-3/padlock.so
/usr/lib64/libcrypto.so.3
/usr/lib64/libcrypto.so.3.2.2
/usr/lib64/libssl.so.3
/usr/lib64/libssl.so.3.2.2
/usr/lib64/ossl-modules
/usr/lib64/ossl-modules/fips.so/usr/lib64/ossl-modules/legacy.so

It does look like the fips provider is being picked up. Strace showed this:

openat(AT_FDCWD, "/usr/lib64/ossl-modules/fips.so", O_RDONLY|O_CLOEXEC) = 7
read(7, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0P\1\0\0\0\0\0"..., 832) = 832
fstat(7, {st_mode=S_IFREG|0755, st_size=2109576, ...}) = 0
mmap(NULL, 2094384, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 7, 0) = 0x7f3278200000
mmap(0x7f3278214000, 1761280, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 7, 0x14000) = 0x7f3278214000
mmap(0x7f32783c2000, 172032, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 7, 0x1c2000) = 0x7f32783c2000
mmap(0x7f32783ec000, 77824, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 7, 0x1eb000) = 0x7f32783ec000
mmap(0x7f32783ff000, 1328, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f32783ff000close(7)

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

**Comment by wgreenberg** [comment]
the plot thickens! if you run `python3 -c 'import ssl; ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)'` within the certbot snap on RHEL 9, do you get the same SSLError?

**Comment by kkeane** [comment]
I am getting the same SSL errror.

**Comment by wgreenberg** [comment]
alright, i was able to setup a RHEL 9-like system with Rocky 9. it has FIPS enabled and uses OpenSSL 3.0.7.

i was indeed able to reproduce the same error, but notably on my setup, python3.12 tries once again to load the FIPS module from `/usr/lib/x86_64-linux-gnu/ossl-modules/fips.so`, which doesn't exist. instead, the module lives at `/usr/lib64/ossl-modules/fips.so`, and symlinking the latter to the former causes certbot to work! so i think this confirms my theory that python3.12's inability to load a FIPS module is what's causing this issue. not sure how we should fix this yet though.

**Comment by kkeane** [comment]
This is an interesting observation! I'm using AlmaLinux 9 as a RHEL-alike system; it should be the same as Rocky here, and in fact the actual path to fips.so is the same.

But notably, strace shows that certbot picks up the fips module from the correct location, yet I'm still getting the error.

openat(AT_FDCWD, "/usr/lib64/ossl-modules/fips.so", O_RDONLY|O_CLOEXEC) = 7
read(7, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0P\1\0\0\0\0\0"..., 832) = 832

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

**Comment by ohemorange** [comment]
Temporary fix for those on ubuntu 22 jammy with fips-updates installed (may work on other systems, haven't tested):

```
sudo mkdir /usr/lib/x86_64-linux-gnu/ossl-modules
sudo ln -s /usr/lib/x86_64-linux-gnu/ossl-modules-3/fips.so /usr/lib/x86_64-linux-gnu/ossl-modules/fips.so
```

**Comment by wgreenberg** [comment]
alright, after endless digging, we found a hacky temporary fix: run certbot with `OPENSSL_FORCE_FIPS_MODE=0`.

this will (as you might expect) force the OpenSSL bundled w/ the snap to ignore the FIPS flag, avoiding this missing provider error and allowing certbot continue to run, albeit without using FIPS-certified functionality. for anyone curious, this flag only appears to be documented in this [debian package patch](https://git.launchpad.net/ubuntu/+source/openssl/tree/debian/patches/fips/crypto-Add-kernel-FIPS-mode-detection.patch?h=applied/ubuntu/noble)

obviously, if you're expecting certbot to use FIPS functions, this won't be sufficient for you. but as mentioned earlier in the thread, even Ubuntu's own FIPS documentation recommends against a snap certbot installation since it's not guaranteed to use the system's FIPS packages. and, indeed, that was never the goal of our snap design. if you want certbot to use your system's FIPS-certified packages, a pip installation is probably what you want for now.

the team's gonna work on what the long-term solution is for now, but i hope this unblocks anyone affected!

**Comment by kkeane** [comment]
I can confirm that a similar, but not identical, workaround works in RHEL 9

mkdir -p /usr/lib/x86_64-linux-gnu/ossl-modules
ln -s /usr/lib64/ossl-modules/fips.so /usr/lib/x86_64-linux-gnu/ossl-modules/fips.so

Two changes to the Ubuntu version:

- mkdir requires the -p argument because the whole directory tree is missing.
- The fips.so module is in a different location.

Thanks everybody for helping out!

Sent with [Proton Mail](https://proton.me/mail/home) secure email.

**Comment by mttjohnson-jf** [comment]
Thank you everyone for helping troubleshoot and understand what was going on here. The [blog article](https://github.com/certbot/certbot/issues/10044#issuecomment-2484180296) that @wgreenberg found above pointing to Canonical recommending FIPS-enabled systems should install certbot via pip was very helpful. Installing with `pip` looks like the best direction going forward for FIPS-enabled systems.

I was able to switch from installing certbot as a snap to now [installing it with pip inside a venv](https://certbot.eff.org/instructions?ws=other&os=pip) on the system, and have v3.0.1 installed and working great now.

I'll stress the importance of setting up the Python virtual environment to run this under for others. At one point I missed a config option and ended up installing `certbot` with `pip` globally on the system, and this ended up breaking `pip` similar to how `certbot` was broken inside the snap where `pip` was complaining about some kind of error in an SSL library or something and it prevented me from running any `pip` commands on the system. Luckily this was all done in the safety of introducing changes on some development virtual infrastructure so I was able to just reset it back to where I started within a couple minutes and make sure I was using `venv` from there on.
